src/connectedk8s/azext_connectedk8s/_clientproxyutils.py

```diff
@@ -0,0 +1,135 @@
+import time
+import azext_connectedk8s._constants as consts
+from base64 import b64encode, b64decode
+from azure.cli.core._profile import Profile
+from azure.cli.core import telemetry
+from azure.cli.core.azclierror import CLIInternalError
+from psutil import process_iter, NoSuchProcess, AccessDenied, ZombieProcess, net_connections
+from knack.log import get_logger
+logger = get_logger(__name__)
+
+
+def check_if_port_is_open(port):
+    try:
+        connections = net_connections(kind='inet')
+        for tup in connections:
+            if int(tup[3][1]) == int(port):
+                return True
+    except Exception as e:
+        telemetry.set_exception(exception=e, fault_type=consts.Port_Check_Fault_Type,
+                                summary='Failed to check if port is in use.')
+        if platform.system() != 'Darwin':
+            logger.info("Failed to check if port is in use. " + str(e))
+        return False
+    return False
+
+
+def close_subprocess_and_raise_cli_error(proc_subprocess, msg):
+    proc_subprocess.terminate()
+    raise CLIInternalError(msg)
+
+
+def check_if_csp_is_running(clientproxy_process):
+    if clientproxy_process.poll() is None:
+        return True
+    else:
+        return False
+
+
+def make_api_call_with_retries(uri, data, method, tls_verify, fault_type, summary, cli_error, clientproxy_process):
+    for i in range(consts.API_CALL_RETRIES):
+        try:
+            response = requests.request(method, uri, json=data, verify=tls_verify)
+            return response
+        except Exception as e:
+            time.sleep(5)
+            if i != consts.API_CALL_RETRIES - 1:
+                pass
+            else:
+                telemetry.set_exception(exception=e, fault_type=fault_type, summary=summary)
+                close_subprocess_and_raise_cli_error(clientproxy_process, cli_error + str(e))
+
+
+def fetch_pop_publickey_kid(api_server_port, clientproxy_process):
+    requestbody = {}
+    poppublickey_uri = f'https://localhost:{api_server_port}/identity/poppublickey'
+    # Needed to prevent skip tls warning from printing to the console
+    original_stderr = sys.stderr
+    f = open(os.devnull, 'w')
+    sys.stderr = f
+
+    get_publickey_response = make_api_call_with_retries(poppublickey_uri, requestbody, "get", False, consts.Get_PublicKey_Info_Fault_Type,
+                                                        'Failed to fetch public key info from clientproxy',
+                                                        "Failed to fetch public key info from client proxy", clientproxy_process)
+
+    sys.stderr = original_stderr
+    publickey_info = json.loads(get_publickey_response.text)
+    kid = publickey_info['publicKey']['kid']
+
+    return kid
+
+
+def fetch_and_post_at_to_csp(cmd, api_server_port, tenantId, kid, clientproxy_process):
+    req_cnfJSON = {"kid": kid, "xms_ksl": "sw"}
+    req_cnf = base64.urlsafe_b64encode(json.dumps(req_cnfJSON).encode('utf-8')).decode('utf-8')
```

Comment:
VM SSH doesn't require `req_cnf` to be base64-encoded:

https://github.com/Azure/azure-cli-extensions/blob/2ce2a17ed87b194ab1088705ba31a2c267d79c50/src/ssh/azext_ssh/custom.py#L187-L193

Is this necessary?

Reply:
This is an example of real-life SSH `data`:

{'token_type': 'ssh-cert', 'req_cnf': '{"kty": "RSA", "n": "ALy4dqBiZKiT3EPSo...", "e": "AQAB", "kid": "5a0ecc0b859302c966b63087c6ade42b44ab52101afb8894d9ba26fd1978318d"}', 'key_id': '5a0ecc0b859302c966b63087c6ade42b44ab52101afb8894d9ba26fd1978318d'}

Reply:
PoP support contract is bit different than ssh cert. We need `req_cnf` to generate PoP token.

Reply:
No. I am not saying VM SSH doesn't require `req_cnf`. It's just VM SSH doesn't require `req_cnf` to be **base64-encoded**.

Reply:
In our PoP contract our Go app expects it to be base64 encoded - https://msazure.visualstudio.com/One/_git/AzureArc-GuestProxy?path=/pkg/adal-with-pop/popkey.go&version=GBmaster&line=96&lineEnd=98&lineStartColumn=1&lineEndColumn=71&lineStyle=plain&_a=contents

Reply:
First of all, congrats to @sirireddy12 and @krdhruva  that this experiment we discussed in January worked out nicely. :-)

I did not notice this PR earlier, and I know it is now merged. Just want to share my 2 cents on this topic.

> > `req_cnf` to be base64-encoded.
> > Is this necessary?
>
> In our PoP contract, [our Go app expects it to be base64 encoded](https://msazure.visualstudio.com/One/_git/AzureArc-GuestProxy?path=/pkg/adal-with-pop/popkey.go&version=GBmaster&line=96&lineEnd=98&lineStartColumn=1&lineEndColumn=71&lineStyle=plain&_a=contents)

The link above is the Go app's implementation which expects the `req_cnf` to be base64-encoded. But do we have that "our PoP contract"? Is the base64 encoding requirement defined in that contract? I got an impression from [this open document](https://datatracker.ietf.org/doc/html/draft-ietf-ace-oauth-params-16#section-3) and its subsequent examples that the `req_cnf` payload does not have base64 encoding.

You don't have to make any changes, if the current base64 encoding behavior is what your internal design requires, and you do not want to adjust it. It is just that the long term maintenance burden would be on the Azure CLI extensions code base: future maintainers would somehow need to know "extension Foo needs base64 encoding, extension Bar does not", and this kind of conversation would provide some context for that future maintainer.

Reply:
This is also my understanding and thoughts. @sirireddy12, I still think it would be better to draft some design spec in the extension source code for future developers or maintainers to understand how "your PoP contract" works.

Reply:
Hey @jiasli / @rayluo , missed this comment. Sure, let me work with my partner team and draft up a one pager about the PoP contract and details